pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java

```diff
@@ -128,26 +128,98 @@ public StudyRoleResponse getStudyRoles() {
-        validateRoleRequest(req);
-
+    public void setUserRoles(SetUserRoleRequest req) {
         int groupId = verifyStudyGroup(studyGroup);
+
+        Map<String, Integer> userIds = validateUsers(req.getUsers(), groupId);
+
         Map<String, RoleInfo> studyRoles = getAdminRoles(groupId);
         List<String> roleNames = req.getRoles();
         validateRoles(roleNames, studyRoles.keySet());
 
-        // pre-check to lessen likelihood of partial operation
-        Map<String, Integer> userIds = new HashMap<>();
-        for (String email: req.getUsers()) {
-            userIds.put(email, verifyUserByEmail(email, groupId).getId());
+        // get existing user roles
+        Map<String, List<String>> userRoles = new HashMap<>();
+        for (var entry: userIds.entrySet()) {
+            userRoles.put(entry.getKey(), getRolesForUser(entry.getValue(), groupId));
         }
 
         for (var entry: userIds.entrySet()) {
-            addRoles(entry.getValue(), roleNames, groupId, studyRoles);
-            log.info("Added roles for user {} in study group {}: {}", entry.getKey(), studyGroup,
-                    String.join(", ", roleNames));
+            List<String> existingRoles = userRoles.get(entry.getKey());
+            Collection<String> rolesToAdd = CollectionUtils.subtract(roleNames, existingRoles);
+            Collection<String> rolesToRemove = CollectionUtils.subtract(existingRoles, roleNames);
+            int userId = entry.getValue();
+
+            if (!rolesToAdd.isEmpty()) {
+                addRoles(userId, rolesToAdd, groupId, studyRoles);
+                log.info("Added roles for user {} in study group {}: {}", entry.getKey(), studyGroup,
+                        String.join(", ", rolesToAdd));
+            }
+
+            if (!rolesToRemove.isEmpty()) {
+                removeUserRoles(userId, rolesToRemove, groupId, studyRoles);
+                log.info("Removed roles for user {} in study group {}: {}", entry.getKey(), studyGroup,
+                        String.join(", ", rolesToRemove));
+            }
+        }
+    }
+
+    /**
+     * Add and remove roles for list of users
+     */
+    public void updateUserRoles(UpdateUserRoleRequest req) {
+        int groupId = verifyStudyGroup(studyGroup);
+
+        if (CollectionUtils.isEmpty(req.getUsers())) {
+            throw new DSMBadRequestException("Invalid users: empty");
+        }
+
+        List<String> addRoles = req.getAddRoles();
+        List<String> removeRoles = req.getRemoveRoles();
+        boolean hasAddRoles = CollectionUtils.isNotEmpty(addRoles);
+        boolean hasRemoveRoles = CollectionUtils.isNotEmpty(removeRoles);
+
+        if (hasAddRoles) {
+            // ensure no union of add and remove roles
+            if (hasRemoveRoles && CollectionUtils.containsAny(addRoles, removeRoles)) {
+                throw new DsmInternalError("Invalid user request: Cannot add and remove the same user");
```

Comment:
Wondering if message can be more clear like: "Cannot add and remove same user  to the same role" ?  OR "Cannot add and remove same role to the same user" ?

Reply:
Ah, good catch... this is not a good message. Will fix.